src/browser/api_protocol/api_handlers/service_middleware.ts

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2017 OpenFin Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { getWindowByUuidName } from '../../core_state';
+import { Identity, OpenFinWindow, ServiceIdentity } from '../../../shapes';
+import { MessagePackage } from '../transport_strategy/api_transport_base';
+import { RemoteAck } from '../transport_strategy/ack';
+import RequestHandler from '../transport_strategy/base_handler';
+import { sendToIdentity } from './api_protocol_base';
+import { Service } from '../../api/service';
+
+const SERVICE_API_ACTION = 'send-service-message';
+const SERVICE_APP_ACTION = 'process-service-action';
+const SERVICE_ACK_ACTION = 'service-ack';
+
+const remoteAckMap: Map<string, RemoteAck> = new Map();
+const pendingServiceConnections: Map<string, MessagePackage[]> = new Map();
+
+function getAckKey(id: number, identity: Identity): string {
+    return `${ id }-${ identity.uuid }`;
+}
+
+function waitForServiceRegistration(uuid: string, msg: MessagePackage) {
+    if (!Array.isArray(pendingServiceConnections.get(uuid))) {
+        pendingServiceConnections.set(uuid, []);
+    }
+    pendingServiceConnections.get(uuid).push(msg);
+}
+
+export function applyPendingServiceConnections(uuid: string) {
+    const pendingConnections = pendingServiceConnections.get(uuid);
+    if (pendingConnections) {
+        pendingConnections.forEach(connectionMsg => {
+            handleServiceApiAction(connectionMsg);
+        });
+        pendingServiceConnections.delete(uuid);
+    }
+}
+
+function setTargetIdentity(identity: Identity, payload: any): { targetIdentity: false | OpenFinWindow, serviceIdentity: ServiceIdentity } {
```

Comment:
Returned shape should be "interfaced"

Reply:
made requested typescript changes @whyn07m3